### DISCUSSION TITLE
Nano Els H4 problem with rotary encoder

### DISCUSSION THREAD

**Discussion by tl1949** [kachurovskiy/nanoels#136] [General]
I haven’t seen any discussion on the following problem so may help if anyone else has the same experience.

When setting up the H4 controller on my lathe, I had a problem as the encoder was not recognised by the controller. Investigation showed that the encoder would only generate a signal as long as the supply voltage did not drop below 5v. The nominal “5v” supply from the H4 controller is around 4.75v and therefore insufficient to drive the encoder. 
Advice from a friend who has more electronics knowledge than me identified that the problem was in the encoder circuitry ie. it has a 7805 voltage regulator.
The problem was resolved by bypassing the 7805 regulator. One leg of the regulator was cut and a wire soldered across the pads on the board. The encoder now generates a signal when powered by the H4 controller and the system works OK.

**Comment by kachurovskiy** [comment]
Hi @tl1949, thank you for sharing this.

There are some regrets that I have w.r.t. PCB design - in particular, I'd like to remove the 2nd fuse (on the encoder 5V line) and lower the resistance of protective resistors on the 5V output lines from 200 Ohm to 50 Ohm to have a better quality control signal on longer wires.

![image](https://github.com/kachurovskiy/nanoels/assets/517919/d0c46c73-c6a0-468e-bde4-fa7dbfeb2738)

These issues can today be worked around by supplying 5V directly by running a 5V wire it into the corresponding screw terminals.

However, I don't think we have a 7805 regulator on the board. Can you please share a picture of your board with the modification that you did?

**Reply by tl1949** [reply]
Hi Maxim,

I should have been clearer, the 7805 is in the encoder. I don't take any credit for the solution, I was advised by a friend who was also informed by an analysis he found on the web at https://domoticx.com/sensor-lpd3806-optical-rotary-encoder/

**Comment by digiexchris** [comment]
On the subject of hardware revisions, If you do another revision, I'd recommend adding a schmidt trigger after the filter in the encoder inputs. Its a modification I did to my H2, and will likely do to my H4 to increase stability.

 I see you use an LC filter. The filter will reduce spikes, but also turn sharp edged square waves into softened edges, more like a sine wave. Noise that hovers near the TTL logic transition point but isn't a high or low enough frequency to be rejected by the filter only needs a tiny fluctuation to flip a LOW to a HIGH, and the frequency doesn't need to be within the filter's cutoff frequency to make it happen. 

The Schmidt trigger uses intentional hysteresis so that the difference between a LOW and a HIGH will never hover near the transition point of the MCU. I've measured this in my own shop on a few similar devices, adding in steps where they shouldn't be, and filter caps helped, but didn't entirely solve it. I ended up going with lower value decoupling caps, and a schmidt trigger. In those projects, other noise mitigation techniques like proper grounding were in place, but some sources could not be fully eliminated. Also, in amateur use, many people won't have the expertise to track that down.

A filter alone, due to the bandwidth widening effect of averaging the noise, will also limit top speed. Combining filter caps with a schmidt trigger can possibly result in much higher pulse rates in drivers that support it.

![image](https://github.com/kachurovskiy/nanoels/assets/5871642/2ef59c2d-09eb-40e8-8fc9-1dda40fac076)

Here's a great writeup on them.
https://www.ti.com/lit/an/scea046a/scea046a.pdf?ts=1701705855935

Not required by any means, but might help out people that work in electrically noisy environments

**Reply by kachurovskiy** [reply]
I don't plan another revision for now. This is cool tech and I considered it but I didn't feel confident that I could get it right and it wasn't strictly required. Also, if we wanted to nail the encoder pulses, Z signal marking a full revolution could be a better investment.

**Comment by digiexchris** [comment]
oh thats a great idea, it could adjust for any lost or added pulses since
the last index pulse, instead of accumulating it! easy to do too, a single
magnet and a hall sensor.

**Reply by kachurovskiy** [reply]
I'm afraid such a sensor wouldn't be precise enough for our purposes. Some encoders just have a separate Z signal that triggers once per turn.

**Comment by duebelpapst** [comment]
Hello everyone,

Yesterday I wanted to cut a thread, but unfortunately I realized that the angle encoder was not working properly.
When I let the chuck make a few turns, there was always a deviation of a few degrees.
Since I had supplied the encoder directly with 5V from the H4, I first changed this and fed external 5.5V directly into the encoder.
Unfortunately this did not help.
The shield of both stepper cables are connected to 5V, only on the H4 side, not at the driver.
So I tried disconnecting the stepper plugs, and it worked, the angle was correct.
Even if connected the X-axis cable only to the H4 and not to the driver, the angle is not correct.
I also tried a different USB cable and a different encoder, but nothing helped.
The strange thing is, 2 weeks ago it was still working.
Does anyone have a tip?

### @Maxim
Further above you wrote:
"These issues can today be worked around by supplying 5V directly by running a 5V wire it into the corresponding screw terminals."
Do you mean that the 5V should be supplied to the encoder AND the H4?

Regards,
Stefan

**Reply by kachurovskiy** [reply]
Hi @duebelpapst, sorry that you're seeing this issue, can you please post the link to the encoder you're using as well as the photos of your encoder and terminals involved? Thank you.

**Reply by duebelpapst** [reply]
Hi,

This problem seems even more difficult than my problem with the VFD / Closed loop JMC.

By the way, the error also occurs when I turn the chuck by hand or with a cordless screwdriver.
If I move the axes while the chuck is turning, the deviation increases.
If I turn the chuck clockwise, I get approx. 1 degree less per 10 revolutions, e.g. 359° instead of 0° etc., if I turn the chuck counterclockwise, I get approx. 1° more per 10 revolutions.
To be honest, this confuses me even more.
If any axis is moved at the same time, the deviation is greater.
As already mentioned, I have already tried two encoders, with 600 p/rev and 360 p/rev, both with the same result. Of course, I had adjusted the ENCODER_STEPS_INT in the software in each case.
I have tried both with 5V from the H4 as well as with external 5.5V.
I also downloaded the software from Github again, just in case I have changed something in the configuration, also without success.
I also laid all the cables separately, i.e. not in a pipe or cable duct.

**Update:**
I have just checked again whether the Z-slide moves correctly during thread cutting, for which I have attached a dial gauge.
The carriage always moves to exactly the same point at both stop points, so the stepper outputs the correct number of pulses, but since the encoder counts incorrectly, it is not synchronized with the rotation.
After the 10 passes with a pitch of 2 mm, a travel of 17.8 mm, at 250 revolutions per minute, only 325° was displayed at my 0° mark. 
So the deviation is less than 1° per 10 revolutions.
That's not much, but it does destroy the thread.
I could now try to compensate for the deviation using a different specification of ENCODER_STEPS_INT, but as it works without the stepper cables, this makes little sense. 
OK I tried 599 instead of 600, and now I get 42° at the 0° mark, which is about right that what I expected.
0,6° x approx. 130 rev = 78°, 325° + 78° = 403° = 360° + 43°


**X-axis plug at the driver side**
![20240128_172033](https://github.com/kachurovskiy/nanoels/assets/45849605/7b36e243-8fef-482f-88a6-a1f61d49e0ed)

**X-axis plug at the H4 side, the shield is connected to +5V**
![20240128_171231](https://github.com/kachurovskiy/nanoels/assets/45849605/ab73bc33-08db-4866-8b15-af162c866909)

**Encoder connection, +5.5V is not connected to the plug, only to the encoder, the shield is connected to GND**
![20240128_171140](https://github.com/kachurovskiy/nanoels/assets/45849605/45df6105-6db0-439e-8616-139d1d186d7e)

**encoder 600 p/rev**
![20240128_171044](https://github.com/kachurovskiy/nanoels/assets/45849605/7ffbd52e-d695-441e-b2ad-dccb35a8aba8)


Regards,
Stefan

**Reply by kachurovskiy** [reply]
@duebelpapst I don't see any obvious problems. I also have some encoder signal loss but it's an order of magnitude less than what you experience.

I can offer you a refund for your H4 controller or I can send you the next iteration of the H4 controller that will have encoder lines feeding more directly into ESP32 inputs, hopefully making it more resistant to noise. The latter will be available in the coming months.

**Reply by kachurovskiy** [reply]
This problem should be resolved in the latest software update (version 12), huge thanks to everyone who helped make it happen ✨

**Comment by jordanpayn1989** [comment]
Hey Maxim,

Finally got my ELS installed after it arrived a few months ago and love it! Everything except threading works really nicely. No stress or anything but thought I'd mention I'm watching for a fix as I have tried 3 different encoders, shielding arrangements and higher voltages fed to encoder and still losing pulses. 

In the photos showing the angles, I've used one of the jaws resting on the toolpost as the 360 degree reference, each successive photo is 1 revolution in the same direction. i have even spun just the encoder well away form anything EMF producing with same results.

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/3990a16a-038e-41f0-9604-b749c943862a)

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/33434745-96f6-4b2d-9d77-2e84573f5688)

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/77529f0e-b63b-4acc-be61-d9e88c7d4481)

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/0fe34898-a7ed-439e-95af-2ac00f7da3d0)

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/70bd7b92-4983-4d26-8e75-6df833cc57ee)

![image](https://github.com/kachurovskiy/nanoels/assets/72195328/b031912f-52b6-47d4-9635-481a03a5a831)

**Reply by kachurovskiy** [reply]
@jordanpayn1989 sorry that you're facing this problem, here are a few things to try:

1. Wiring the terminal as close to the shielding as possible ![image](https://github.com/kachurovskiy/nanoels/assets/517919/42a57e2b-6cb5-451b-a74d-ba96a2fa18f6)

2. Changing your `spinEnc()` method to ignore very short pulses

```
void IRAM_ATTR spinEnc() {
  static uint32_t lastTime = 0;
  uint32_t currentTime = micros();
  if (currentTime - lastTime < 10) return;
  lastTime = currentTime;
  spindlePosDelta += digitalRead(ENC_B) ? -1 : 1;
}
```

3. Using the h4.ino version where we listen to encoder pulses with hardware pulse counter - https://gist.github.com/kachurovskiy/e8842a36b95ad179654dc50b461511ce

**Reply by jordanpayn1989** [reply]
Thanks heaps Maxim!

1. Shortened wiring - Didn't work

2. Looked into but saw someone mention hardware counting and thought it would be a better bet so looked at your version 12

3. Absolutely dead on perfect now! Thanks for your continued RnD on this. 

It is so amazing being able to change pitch on the fly. Id always groan when change gears needed adjusting and would just put the job off. 

Very happy Ausie!
